### DISCUSSION TITLE
ERROR: Key @PLAINNAME@ is not in the dictionary

### DISCUSSION THREAD

**Discussion by stsp** [mesonbuild/meson#12595] [Q&A] [answered]
Hi, I am trying to write this generator:
```
gtgen = generator(find_program(PD),
  arguments: ['@INPUT@', SRC / 'proto.h'],
  output: GEN_TMP['@PLAINNAME@'],
  capture: true)
```

And what I get, is:
`../meson.build:76:18: ERROR: Key @PLAINNAME@ is not in the dictionary.`
Why the key wasn't expanded before a
dictionary look-up?

**Comment by eli-schwartz** [comment]
Dictionaries are parsed by the interpreter first, since they are used as arguments to the function.  The return value of the dictionary lookup is used as an input argument to the function.

This is how most programming languages work.

**Reply by eli-schwartz** [reply]
As a matter of curiosity, why do you want to use a different template for each output file? The purpose of a generator is to have a distinct input->output correspondence.

**Reply by stsp** [reply, marked as answer]
I am generating lots of unrelated files.
There is a shell script that generates
them by the number. I.e.
`gen.sh 100` generates the file number
100 and dumps it to stdout, and the name
is assigned to it via dictionary.
This was done to improve parallel builds.
I can of course generate all files in 1 go.
But they go to entirely unrelated build
target chains, so I made them generated
on demand to not produce a stall point
for parallel builds.

**Reply by eli-schwartz** [reply]
A generator is very similar to a custom_target run inside of a `foreach` loop.

The generator sacrifices flexibility (you can't specify as many details) in order to be simpler to use (you don't *have* to specify as many details).

**Comment by stsp** [comment]
So what would you suggest?
Ideally, I'd like to be able to generate
many custom_target()s with a foreach loop. :)
But I don't suppose meson supports
such a meta-programming tricks.

So I am right that I only have 2 solutions
here: either copy/paste custom_target()s
and adjust each by hands, or generate all
files in 1 go, adding the stall point for
parallel builds.

Which of the above would you suggest,
or maybe you can think of something else?

**Comment by stsp** [comment]
I tried some work-arounds, but the
problem seems to be that meson
insists on the generator's input being
an actual existing file.
How can I disable such a check?
I'd like to pass arbitrary args to a
generator. Why meson prevents me
from doing that?

**Comment by stsp** [comment]
OK, I solved the problems by
tricking around the `extra_args`
property, together with writing
out some targets manually.
